subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/report/HtmlDependencyVerificationReportRenderer.java

```diff
@@ -0,0 +1,442 @@
+    private Section currentSection;
+    private final StringBuilder contents = new StringBuilder();
+    private final DocumentationRegistry documentationRegistry;
+    private final File verificationFile;
+    private final List<String> writeFlags;
+    private final File htmlReportOutputDirectory;
+
+    HtmlDependencyVerificationReportRenderer(DocumentationRegistry documentationRegistry, File verificationFile, List<String> writeFlags, File htmlReportOutputDirectory) {
+        this.documentationRegistry = documentationRegistry;
+        this.verificationFile = verificationFile;
+        this.writeFlags = writeFlags;
+        this.htmlReportOutputDirectory = htmlReportOutputDirectory;
+    }
+
+    @Override
+    public void title(String title) {
+        currentSection = new Section(title);
+        this.sections.add(currentSection);
+    }
+
+    @Override
+    public void withErrors(Runnable action) {
+        action.run();
+    }
+
+    @Override
+    public void withArtifact(ModuleComponentArtifactIdentifier key, Runnable action) {
+        currentSection.newArtifact(new ArtifactErrors(key));
+        action.run();
+
+    }
+
+    @Override
+    public void reportFailure(RepositoryAwareVerificationFailure failure) {
+        currentSection.currentArtifact.addFailure(failure);
+    }
+
+    @Override
+    public void multipleErrors(Runnable action) {
+        action.run();
+    }
+
+    @Override
+    public void finish(DependencyVerificationReportWriter.HighLevelErrors highLevelErrors) {
+
+    }
+
+    public void renderNavBar() {
+        contents.append("<nav class=\"uk-navbar-container\" uk-navbar>\n" +
+            "    <div class=\"uk-navbar-left\">\n" +
+            "        <a href=\"\" class=\"uk-navbar-item uk-logo\"><img src=\"img/gradle-logo.png\" width=\"120\"></a>\n" +
+            "        <ul class=\"uk-navbar-nav\">\n" +
+            "<li class=\"uk-active\"><a href=\"#\">Dependency verification report</a></li>\n" +
+            "        </ul>\n" +
+            "    </div>\n" +
+            "</nav>\n" +
+            "\n");
+    }
+
+    public void renderSections() {
+        contents.append("<div class=\"uk-container uk-container-expand\">\n");
+        contents.append("        <ul uk-accordion>\n");
+        boolean first = true;
+        for (Section section : sections) {
+            if (first) {
+                contents.append("            <li class=\"uk-open\">\n");
+            } else {
+                contents.append("            <li>\n");
+            }
+            prerenderSection(section);
+            renderSection(section);
+            contents.append("            </li>\n");
+            first = false;
+        }
+        contents.append("         </ul>\n");
+        contents.append("        </div>\n");
+    }
+
+    File writeReport() {
+        generateContent();
+        ensureReportDirectoryCreated();
+        copyReportResources();
+        return doWriteReport();
```

Comment:
Don't we have the proper infrastructure to build up the report incrementally with title()/finish()?

Reply:
A report is _written to disk_ everytime a verification failure happens. By verification failure, I mean in general a graph resolution fails with one or more errors. So this method is called several times, but _only_ in the "lenient" case where we actually build up from previous failed resolutions. Not ideal, but it offers a couple of advantages:

1. you can start reading the reports as verification failures happen
2. it's `CTRL+C` safe: once you see a failure on the console, you can actually open the report right away and interrupt the build without risking to lose this information.